### PR TITLE
mmapstorage: store the OPMPHM

### PR DESCRIPTION
**Pull request by mpranj**
Store the OPMPHM and generally improve mmapstorage (some critical fixes and sanity checks).
Disable `testmod_zeromqsend`.
Increase Jenkins CI timeout #3157.

## Basics

These points need to be fulfilled for every PR:

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy.
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [x] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [x] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

**Comment by mpranj** [comment]
[Build failed](https://build.libelektra.org/jenkins/blue/organizations/jenkins/libelektra/detail/PR-3136/2/pipeline) because of docker: 
Record trace of a Docker image used in FROM: `remote file operation failed`.

I don't know if this is a known problem but seems unrelated to the PR.

**Comment by markus2330** [comment]
Thank you for creating this PR! Good to see activity from you :sparkling_heart:

**Comment by mpranj** [comment]
@markus2330 I ran into #2130 once again. Since the addition of OPMPHM to mmap, I'd need the `opmphmDel` and `opmphmPredictorDel` functions to free the old resources before replacing them by mmap'ed ones.

The implementations are "entangled" either way, since mmapstorage needs knowledge of any structures it stores. The alternative to exporting the symbols is to copypaste the `*Del` functions into mmapstorage. On the positive side: it's really not too much code, and any change to the core structs (Key, KeySet, OPMPHM) inevitably implies adapting mmapstorage accordingly.

What approach to you like more: exporting vs copypasta?

**Comment by mpranj** [comment]
Now the jenkins builds have finished successfully, but the status is not reported to github. 🤷‍♂

**Comment by sanssecours** [comment]
> Now the jenkins builds have finished successfully, but the status is not reported to github. 🤷‍♂

That happened in one of my PRs (a long time ago) too. I just restarted the build when it happened. For obvious reasons – Jenkins build fails all the time and takes forever – this is not an ideal solution. Anyway, I would consider this PR “ready to merge”.

**Comment by markus2330** [comment]
In some cases it is unfortunately necessary to open a new PR. Jenkins build are now hopefully faster as we now got one more server.

**Comment by mpranj** [comment]
jenkins build libelektra please

**Comment by mpranj** [comment]
jenkins build libelektra please

**Comment by mpranj** [comment]
jenkins build libelektra please

**Comment by mpranj** [comment]
jenkins build libelektra please

**Comment by mpranj** [comment]
@markus2330 this should be done, but the tests are constantly failing because zeromq or gopts tests are timing out.

**Comment by markus2330** [comment]
The zeromq tests can be disabled (as nobody is working on this). Can you please report the gopts issue, maybe @kodebach finds a solution.

**Comment by mpranj** [comment]
jenkins build libelektra please

**Comment by markus2330** [comment]
Please document in doc/todo/TESTING which tests were disabled (ideally also add commit SHA-1 how to revert the disabling of tests).

**Comment by mpranj** [comment]
Seems like `testmod_zeromqsend` was disabled, but not for `kdb run_all`. I will disable it there now, instead of removing the tests completely. The removal was already documented, but only partially implemented.

**Comment by mpranj** [comment]
@markus2330 done.

File: src/libs/elektra/opmphm.c [hunk 1]
```diff
@@ -13,6 +13,7 @@
 #include <kdblogger.h>
 #include <kdbmacros.h>
 #include <kdbopmphm.h>
+#include <kdbprivate.h>
```
Comment (markus2330) [inline src/libs/elektra/opmphm.c]: Please add the reason of this include as comment next to it.

File: src/libs/elektra/opmphmpredictor.c [hunk 2]
```diff
@@ -10,6 +10,7 @@
 #include <kdbhelper.h>
 #include <kdblogger.h>
 #include <kdbopmphmpredictor.h>
+#include <kdbprivate.h>
```
Comment (markus2330) [inline src/libs/elektra/opmphmpredictor.c]: Please add the reason of this include as comment next to it.

File: src/plugins/mmapstorage/internal.h [hunk 3]
```diff
@@ -99,7 +124,12 @@ typedef struct _mmapFooter MmapFooter;
 /**
  * Mmap information header
  *
- * shall contain only fixed-width types
+ * Changes to this struct can/will break compatibility with existing files.
+ * If a breaking changed needs to be introduced, change `ELEKTRA_MAGIC_MMAP_NUMBER`
```
Comment (markus2330) [inline src/plugins/mmapstorage/internal.h]: Please give details what a breaking change is:

- reordering of struct?
- adding to struct?
- ...

**Comment by markus2330** [comment]
Thank you, great work! I am looking forward to see the benchmarks!

**Pull request merged**